Add project toolchain documentation (#32901)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/platforms/jvm/toolchains.adoc

====
[.multi-language-sample]
=====
[source,kotlin]
----
tasks.withType<JavaCompile>().configureEach {
    options.release = 8
}
----
=====
[.multi-language-sample]
=====
[source,groovy]
----
tasks.withType(JavaCompile).configureEach {
    options.release = 8
}
----
=====
====

This flag prevents accidental use of newer APIs that are not available in the specified version.
However, it does not control which JDK is used—only how the compiler treats source code.

This method can be combined with toolchains **if you need both a specific JDK and strict cross-compilation**.

[[sec:source-target-toolchain]]
=== 3. Source and Target compatibility

Setting `sourceCompatibility` and `targetCompatibility` tells the Java compiler to produce bytecode compatible with a specific Java version but does *not* enforce which JDK Gradle itself runs with:

====
[.multi-language-sample]
=====
[source,kotlin]
----
java {
    sourceCompatibility = JavaVersion.VERSION_1_8
    targetCompatibility = JavaVersion.VERSION_1_8
}
----
=====
[.multi-language-sample]
=====
[source,groovy]
----
java {
    sourceCompatibility = JavaVersion.VERSION_1_8
    targetCompatibility = JavaVersion.VERSION_1_8

> }

Reviewer: 💭  Does happen often that Kotlin and Groovy snippets are identical? Would it make sense to add a feature to our multi-language-sample thing to display the sample in both languages, but specify it only once in the source?

----
=====
====

This does *not* guarantee the correct JDK is used and may cause issues when APIs have been backported to older Java versions.

You should only use this method in cases where you need backward compatibility **but cannot use toolchains**.

[[sec:java-home-toolchain]]
=== 4. Environment variables (`JAVA_HOME`)

You can influence which JDK Gradle uses by setting the `JAVA_HOME` environment variable:

[source,bash]
----
export JAVA_HOME=/path/to/java17
----

This sets a default JDK for all Java-based tools on your system, including Gradle and Maven.

WARNING: This does not override Gradle’s toolchain support or other project-specific configurations.

This approach is useful for legacy projects that do not use toolchains and expect a specific JDK to be active in the environment.

However, since `JAVA_HOME` applies globally, it cannot be used to specify different JDK versions for different projects.
It is more reliable to use <<sec:java-toolchains,toolchains>>, which allow setting the Java version at the project level.

[[sec:ide-settings-toolchain]]

> === 5. IDE settings

Reviewer: ❓  Should this section mention the Daemon Toolchains? IIRC, they should be picked up by the IDE automatically. Though, I am not sure if the support for this has already landed in a public IDE releases.

Author: I saw this in the IntelliJ IDEA EAP. Not sure if it made it in the main release yet. I will still add a note.

Most modern IDEs allow you to specify the Java version for Gradle independently.

> This can be useful if you want Gradle to match the JDK used in your development environment.

Reviewer: 🤔  After re-reading this, I am not sure I understand what this is trying to say.

The IDE setting controls the JVM that would be used to run Gradle in the given project. This can affect the Java version with which the project code is compiled, if the project does not specify anything about toolchains (aka does not use any of the other options described on this page). This is very undesirable, and should never be the case, because it hurts reproducibility.

In my understanding, changing this setting mostly makes sense to ensure that the Java version running Gradle on the command line matches the one used to run Gradle via the IDE. E.g. when you run tests via the IDE menus.

==== IntelliJ IDEA

To configure the Gradle JVM:

1. Open *Settings (Preferences)* > *Build, Execution, Deployment* > *Gradle*.
2. Set *Gradle JVM* to the desired JDK.

==== Eclipse

To configure the Gradle JVM:

1. Open *Preferences* > *Gradle* > *Gradle JDK*.
2. Select the appropriate JDK.

=== Combining toolchains

In some cases, you may want to:

- Use a **specific JDK version** for compilation (`toolchains`).
- Ensure that the compiled bytecode is **compatible with an older Java version** (`--release` or `targetCompatibility`).

For example, to compile with Java 17 but produce Java 11 bytecode:

====
[.multi-language-sample]
=====
[source,kotlin]
----
java {
    toolchain {

>         languageVersion.set(JavaLanguageVersion.of(17))

Reviewer: ```suggestion
        languageVersion = JavaLanguageVersion.of(17)
```

Author: Missed this one. TY

    }
}

tasks.withType<JavaCompile>().configureEach {
    options.release = 11
}
----
=====
[.multi-language-sample]
=====
[source,groovy]
----
java {
    toolchain {
        languageVersion = JavaLanguageVersion.of(17)
    }
}

tasks.withType(JavaCompile).configureEach {
    options.release = 11
}
----
=====
====

=== Comparison table for setting project toolchains

|===
| Method | Ensures Correct JDK? | Auto Downloads JDK? | Prevents Accidental API Use?

| Java toolchains
| ✅ Yes
| ✅ Yes
| ❌ No

| `--release` flag
| ❌ No
| ❌ No
| ✅ Yes

| Source & Target compatibility
| ❌ No
| ❌ No
| ❌ No

| Environment variables (`JAVA_HOME`)
| ✅ Yes (but only globally)
| ❌ No
| ❌ No

| IDE settings
| ✅ Yes (inside the IDE)
| ❌ No
| ❌ No
|===

Recommendation:

- *For most users:* Use Java toolchains (`toolchain.languageVersion`).
- *For strict compatibility enforcement:* Use the `--release` flag.
- *For advanced cases:* Combine toolchains and `--release`.
- *Avoid* `sourceCompatibility` and `targetCompatibility` unless necessary.

> - *Use `JAVA_HOME`* only if you need to enforce a system-wide JDK version.

Reviewer: ```suggestion
- *Use `JAVA_HOME`* only if you need a default system-wide JDK version.
```

Usage of _invalid_ instances of `JavaToolchainSpec` results in a build error since Gradle 8.0.

[[sec:source-target-toolchain]]
=== 2. Source and Target compatibility

Setting `sourceCompatibility` and `targetCompatibility` tells the Java compiler to produce bytecode compatible with a specific Java version but does *not* enforce which JDK Gradle itself runs with:

> [source,kotlin]

Reviewer: 🤔  Only the newly added snippets seem to not have the Kotlin + Groovy version. Should we make it consistent across the page and add Groovy?

----
java {
    sourceCompatibility = JavaVersion.VERSION_1_8
    targetCompatibility = JavaVersion.VERSION_1_8
}
----

This does *not* guarantee the correct JDK is used and may cause issues when APIs have been backported to older Java versions.

You should only use this method in cases where you need backward compatibility **but cannot use toolchains**.

[[sec:flag-toolchain]]

> === 3. The `--release` flag

Reviewer: 💭  By the time I scrolled until here, I am lost. I don't know what "list" this third item refers to.
The first toolchains sub-section is very long and the small difference in the heading size does not really help recovering the context. I could have consulted the page structure (rendered on the right), but it only shows top-level stuff, so it's not helpful to orient myself.

We don't have to do anything about it in this PR, but it's a general concern worth thinking over.

For strict cross-compilation, the `--release` flag is recommended instead of `sourceCompatibility` and `targetCompatibility`:

[source,kotlin]
----
tasks.withType<JavaCompile> {

>     options.release.set(8)

Reviewer: Why not assignment?

```suggestion
    options.release = 8
```

}
----

This flag prevents accidental use of newer APIs that are not available in the specified version.
However, it does not control which JDK is used—only how the compiler treats source code.

This method can be combined with toolchains **if you need both a specific JDK and strict cross-compilation**.

[[sec:java-home-toolchain]]
=== 4. Environment variables (`JAVA_HOME`)

Instead of using Gradle toolchains, you can manually specify a JDK globally using the `JAVA_HOME` environment variable:

[source,bash]
----
export JAVA_HOME=/path/to/java17
----

This approach ensures that all Java-based tools (Gradle, Maven, etc.) use the same JDK version.
It enforces a system-wide JDK version and is great for legacy projects that do not support toolchains.

However, it applies the Java version at a global scope, so it is not possible to specify different JDKs per project.

[[sec:ide-settings-toolchain]]
=== 5. IDE settings

Most modern IDEs allow you to specify the Java version for Gradle independently.
This can be useful if you want Gradle to match the JDK used in your development environment.

==== IntelliJ IDEA

To configure the Gradle JVM:

> 1. Open *Settings (Preferences)* > *Build, Execution, Deployment* > *Gradle*.

Author: I've only tried it on code reviews. I can try it on doc reviews next time.
